### PR TITLE
Cartesian product data structure with appropriate functions is added

### PR DESCRIPTION
**Pull request by dmitry-vlasov**

### CODE REVIEW

File: lib/ds/prod.flow [hunk 1]
```diff
@@ -0,0 +1,114 @@
+import ds/array;
+
+// Cartesian product of arrays: if c = [a_1, a_2, ... , a_n] then
+// c is interpreted as a_1 x a_2 x ... x a_n
+
+export {
+	// Cartesian product of sets (elements of the array).
+	Prod : (sets : [[?]]);
```
Comment (alstrup) [inline lib/ds/prod.flow]: Too short a name. Make it longer.

File: lib/ds/prod.flow [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+	isEmptyProd(prod : Prod<?>) -> bool;
+
+	// Map all tuples of product. The result has sizeProd(prod) elements.
+	mapProd(prod : Prod<?>, f : ([?]) -> ??) -> [??];
+
+	// Traverse all tuples of product.
+	iterProd(prod : Prod<?>, f : ([?]) -> void) -> void;
+	foldProd(prod : Prod<?>, init : ??, f : (??, [?]) -> ??) -> ??;
+	fltermapProd(prod : Prod<?>, f : ([?]) -> Maybe<??>) -> [??];
```
Comment (alstrup) [inline lib/ds/prod.flow]: filter with i

File: lib/ds/prod.flow (hunk 1 above)
Comment (dmitry-vlasov) [reply lib/ds/prod.flow]: Renamed to CartProd

File: lib/ds/prod.flow (hunk 2 above)
Comment (dmitry-vlasov) [reply lib/ds/prod.flow]: fixed

**Pull request merged**